lms/djangoapps/experiments/views_custom.py

```diff
@@ -0,0 +1,166 @@
+    # http://localhost:18000/api/experiments/v0/custom/REV-934/?course_id=course-v1:edX+DemoX+Demo_Course
+
+    authentication_classes = (
+        JwtAuthentication,
+        OAuth2AuthenticationAllowInactiveUser,
+        SessionAuthenticationAllowInactiveUser,
+    )
+    permission_classes = (ApiKeyHeaderPermissionIsAuthenticated,)
+
+    @method_decorator(ensure_csrf_cookie_cross_domain)
+    def get(self, request):
+        """
+        Return the if the course should be upsold in the mobile app, if the user has appropriate permissions.
+        """
+        if not MOBILE_UPSELL_FLAG.is_enabled():
+            return Response({
+                'show_upsell': False,
+                'upsell_flag': False,
+            })
+
+        if not 'course_id' in request.GET:
+            return Response({
+                'show_upsell': False,
+            })
+
+        course_id = request.GET.get('course_id').replace(' ', '+')  # HACK: the url decoding converts plus to space; put them back
+        course_key = CourseKey.from_string(course_id)
+        course = CourseOverview.get_from_id(course_key)
+        user = request.user
+
+        enrollment = None
+        user_enrollments = None
+        audit_enrollments = None
+        has_non_audit_enrollments = False
+        try:
+            user_enrollments = CourseEnrollment.objects.select_related('course').filter(user_id=user.id)
+            has_non_audit_enrollments = user_enrollments.exclude(mode__in=CourseMode.UPSELL_TO_VERIFIED_MODES).exists()
+            enrollment = CourseEnrollment.objects.select_related(
+                'course'
+            ).get(user_id=user.id, course_id=course.id)
+        except CourseEnrollment.DoesNotExist:
+            pass  # Not enrolled, use the default values
+
+        context = get_base_experiment_metadata_context(course, user, enrollment, user_enrollments)
+
+        bucket = stable_bucketing_hash_group(MOBILE_UPSELL_EXPERIMENT, 2, user.username)
+        if hasattr(request, 'session') and MOBILE_UPSELL_EXPERIMENT not in request.session:
+            properties = {
+                'site': request.site.domain,
+                'app_label': 'experiments',
+                'bucket': bucket,
+                'experiment': 'REV-934',
+            }
+            segment.track(
+                user_id=user.id,
+                event_name='edx.bi.experiment.user.bucketed',
+                properties=properties,
+            )
+
+            # Mark that we've recorded this bucketing, so that we don't do it again this session
+            request.session[MOBILE_UPSELL_EXPERIMENT] = True
+
+        show_upsell = bucket != 0 and not has_non_audit_enrollments
+        if show_upsell:
+            return Response({
+                'show_upsell': show_upsell,
+                'price': context.get('upgrade_price'),
+                'basket_url': context.get('upgrade_link'),
+            })
+        else:
+            return Response({'show_upsell': show_upsell})
```

Comment:
Maybe add the flag here? the flag is on, but there is still no upsell

Reply:
Bucket also seems useful, I'll add both